addons/purchase_stock/models/stock.py

```diff
@@ -19,6 +19,42 @@ class StockMove(models.Model):
     created_purchase_line_id = fields.Many2one('purchase.order.line',
         'Created Purchase Order Line', ondelete='set null', readonly=True, copy=False)
 
+    def _action_cancel_origin(self):
+        """ Update purchase order in case of draft state.
+            > Check if all previouse moves cancelled , then cancel or unlink purchase line.
+            > Check if some of previouse moves are not cancel decrease the line quantity.
+            > Check that purchase line has been generated from the multiple move or not.
+        """
+        moves_purchase_line = self.filtered(lambda m: m.created_purchase_line_id and m.state not in ('done', 'cancel'))
+        for move in moves_purchase_line:
+            order = move.created_purchase_line_id.order_id
+            purchase_line = move.created_purchase_line_id
+            if move.previous_move_propagate:
+                if order.state not in ['done', 'purchase']:
+                    siblings_states = (move.created_purchase_line_id.move_dest_ids - move).mapped('state')
+                    if all(state == 'cancel' for state in siblings_states):
+                        if len(order.order_line) > 1:
+                            purchase_line.unlink()
```

Comment:
it is generally a bad idea to unlink inside a loop, you'll break the cache

Reply:
changes done